plugins/outputs/whatap/whatap.go

```diff
@@ -0,0 +1,382 @@
+package whatap
+
+import (
+	"fmt"
+	"log"
+	"net"
+	"os"
+	"sort"
+	"strings"
+	"time"
+
+	"github.com/influxdata/telegraf"
+
+	//"github.com/influxdata/telegraf/internal"
+	"github.com/influxdata/telegraf/plugins/outputs"
+)
+
+const (
+	NETSRC_AGENT_ONEWAY = 10
+	VALUE_DECIMAL       = 20
+	VALUE_DOUBLE        = 40
+	VALUE_TEXT          = 50
+	VALUE_MAP           = 80
+	// 5633
+	PACK_TAG_COUNT = 0x1601
+)
+
+type Whatap struct {
+	License string        `toml:"license"`
+	Servers []string      `toml:"servers"`
+	Pcode   int64         `toml:"pcode"`
+	Timeout time.Duration `toml:"timeout"`
+	Oname   string
+	Oid     int32
+	Session TcpSession
+}
+
+type TcpSession struct {
+	Client net.Conn
+	Dest   int
+}
+
+var sampleConfig = `
+  ## You can create a project on the WhaTap site(https://www.whatap.io) 
+  ## to get license, project code and server IP information.
+
+  ## WhaTap license. Required
+  #license = "xxxx-xxxx-xxxx"
+
+  ## WhaTap project code. Required
+  #pcode = 1111
+
+  ## WhaTap server IP. Required
+  ## Put multiple IPs. ["tcp://1.1.1.1:6600","tcp://2.2.2.2:6600"]
+  #servers = ["tcp://1.1.1.1:6600","tcp://2.2.2.2:6600"]
+
+  ## Connection timeout.
+  # timeout = "60s"
+`
+
+func (w *Whatap) Connect() error {
+	//log.Println("[outputs.whatap] Connect")
+	w.Session.Dest += 1
+	if w.Session.Dest >= len(w.Servers) {
+		w.Session.Dest = 0
+	}
+
+	addr := strings.SplitN(w.Servers[w.Session.Dest], "://", 2)
+	if len(addr) != 2 {
+		return fmt.Errorf("invalid address: %s", w.Servers[w.Session.Dest])
+	}
+	if addr[0] != "tcp" {
+		return fmt.Errorf("only tcp is supported: %s", w.Servers[w.Session.Dest])
+	}
+
+	t := w.Timeout * time.Millisecond
+	if client, err := net.DialTimeout(addr[0], addr[1], t); err != nil {
+		return err
+	} else {
+		w.Session.Client = client.(*net.TCPConn)
+		log.Println("I! [outputs.whatap] Connected tcp to ", addr)
+	}
+
+	return nil
+
+}
+
+func (w *Whatap) Close() error {
+	if w.Session.Client == nil {
+		return nil
+	}
+	err := w.Session.Client.Close()
+	w.Session.Client = nil
+	return err
+}
+
+func (w *Whatap) Description() string {
+	return "Configuration for WhaTap"
+}
+
+func (w *Whatap) SampleConfig() string {
+	return sampleConfig
+}
+
+func (w *Whatap) Write(metrics []telegraf.Metric) error {
+	if len(metrics) == 0 {
+		return nil
+	}
+	if w.Session.Client == nil {
+		if err := w.Connect(); err != nil {
+			w.Close()
+			return err
+		}
+	}
+	for _, m := range metrics {
+		p := NewTagCountPack()
+		p.Pcode = w.Pcode
+		p.Oid = w.Oid
+		p.Category = fmt.Sprintf("%s%s", "telegraf_", m.Name())
```

Comment:
You could easily setup the struct directly here
```suggestion
		p := TagCountPack {
			Pcode: w.Pcode,
			Oid: w.Oid,
			Category: "telegraf_" + m.Name(),
			Tags: make(map[string]string),
			Data: make(map[string]interface{}),
		}
```

Reply:
There is an initialization process inside the external library NewTagCountPack() function, so it is used without modification.